api/agent/timer_desired_state_feedback_test.go

```diff
@@ -0,0 +1,322 @@
+		mockClient.EXPECT().SendDesiredStateFeedback(gomock.Any(), gomock.Any()).Times(1)
+		notifier.set(testActivityID, actions)
+		assert.Equal(t, actions, notifier.actions)
+		assert.Equal(t, reportedActions, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		time.Sleep(1 * time.Second)
+		mockClient.EXPECT().SendDesiredStateFeedback(gomock.Any(), gomock.Any()).Times(1)
+		notifier.set(testActivityID, actions2)
+		assert.Equal(t, actions2, notifier.actions)
+		assert.NotEqual(t, reportedActions2, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		time.Sleep(2 * time.Second)
+		assert.Equal(t, actions2, notifier.actions)
+		assert.Equal(t, reportedActions2, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		if notifier.internalTimer != nil {
+			if !notifier.internalTimer.Stop() {
+				<-notifier.internalTimer.C
+			}
+		}
+	})
+	t.Run("test_no_event_published_on_resetting_the_same_actions_during_timeout", func(t *testing.T) {
+		mockClient := mocks.NewMockUpdateAgentClient(mockCtr)
+		interval := 2 * time.Second
+		updAgent := &updateAgent{
+			client: mockClient,
+		}
+		notifier := newDesiredStateFeedbackNotifier(interval, updAgent)
+		mockClient.EXPECT().SendDesiredStateFeedback(gomock.Any(), gomock.Any()).Times(1)
+		notifier.set(testActivityID, actions)
+		assert.Equal(t, actions, notifier.actions)
+		assert.Equal(t, reportedActions, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		time.Sleep(1 * time.Second)
+		mockClient.EXPECT().SendDesiredStateFeedback(gomock.Any(), gomock.Any()).Times(0)
+		notifier.set(testActivityID, actions)
+		time.Sleep(2 * time.Second)
+		assert.Equal(t, actions, notifier.actions)
+		assert.Equal(t, reportedActions, notifier.reportedActions)
+		assert.Equal(t, testActivityID, notifier.activityID)
+		if notifier.internalTimer != nil {
+			notifier.internalTimer.Stop()
+		}
+	})
+}
+
+func TestDesiredStateFeedbackTimerStop(t *testing.T) {
+	mockCtr := gomock.NewController(t)
+	defer mockCtr.Finish()
+
+	t.Run("test_stop", func(t *testing.T) {
+		mockClient := mocks.NewMockUpdateAgentClient(mockCtr)
+		interval := time.Second
+		updAgent := &updateAgent{
+			client: mockClient,
+		}
+
+		actions := []*types.Action{
+			{
+				Component: &types.Component{
+					ID:      "action 1 component",
+					Version: "1.0.0",
+				},
+				Status:  types.ActionStatusUpdateSuccess,
+				Message: "update success",
+			},
+		}
+		reportedActions := []*types.Action{
+			{
+				Component: &types.Component{
+					ID:      "action 1 component",
+					Version: "1.0.0",
+				},
+				Status:  types.ActionStatusUpdateSuccess,
+				Message: "update success",
+			},
+		}
+		notifier := newDesiredStateFeedbackNotifier(interval, updAgent)
+		notifier.actions = actions
+		notifier.reportedActions = reportedActions
+		notifier.internalTimer = time.AfterFunc(interval, func() {})
```

Comment:
the notifier initialization is repeated trough the test cases a lot, please consider making and `init` func for it